### PR TITLE
CAM-13699: feat(spin): strip empty lines

### PR DESCRIPTION
**Pull request by jomu78**
- avoid empty lines if an already pretty printed XML is inserted into a SpinXml Element

fixes CAM-13699

### CODE REVIEW

**Comment by koevskinikola** [comment]
Hi @jomu78,

Thank you for this contribution. It definitely looks interesting. I will go through the Jira ticket you created, as well as the code, and provide you with feedback asap.

Best,
Nikola

**Review by koevskinikola — CHANGES_REQUESTED** [review]
I have some suggestions to make the new code in line with our coding style guide.

File: dataformat-xml-dom/src/test/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriterTest.java
Comment (koevskinikola) [inline dataformat-xml-dom/src/test/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriterTest.java:18]: A test case that reproduces the existing behavior would also be useful here, as we would be able to assert the behavior for both with pretty print and without it.

**Review by koevskinikola — CHANGES_REQUESTED** [review]

**Comment by jomu78** [comment]
@koevskinikola 
I have updated the PR and I hope all requested changes have been taken into account.

**Comment by jomu78** [comment]
@koevskinikola : Did you have some time to check the changes I have made? Are there further changes required to accept the PR?

File: dataformat-xml-dom/src/main/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormat.java
Comment (koevskinikola) [inline dataformat-xml-dom/src/main/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormat.java:56]: Setting the `prettyPrint` property to true would break backward compatibility. Any users that currently expect the output to have an empty line would face unexpected behaviors.

We should also keep with the pattern of setting property values in the constructor.

I would remove the value here:
```suggestion
  protected boolean prettyPrint;
```

And adjust the constructor on lines 75-80 to set the `prettyPrint` property. By default, `false` should be passed.

File: dataformat-xml-dom/src/test/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriterTest.java
Comment (koevskinikola) [inline dataformat-xml-dom/src/test/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriterTest.java:18]: As mentioned above, the pretty print feature should be disabled by default. Otherwise, we will introduce some unexpected behavior to users. The tests here should be adjusted accordingly.

**Review by koevskinikola — CHANGES_REQUESTED** [review]
I have some additional review hints. 

However, when reviewing the code, I realized that it's not exactly clear why the empty lines are introduced in the first place. In CAM-13699, you wrote:

> The XML is formatted again and any CR returns an empty line. When reading the XML and then inserting it again any blank line is even doubled. 

It's still not clear to me what actually doubles the blank lines? Can you provide some information on this?

Otherwise, your contribution just fixes the symptoms, but the bug is still there.

File: dataformat-xml-dom/src/main/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormat.java
Comment (jomu78) [reply dataformat-xml-dom/src/main/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormat.java:56]: in the current master branch DomXmlDataFormatWriter#getTransformer returns a transformer with property INDENT set - so with default setting it DOES formatting. With prettyPrint = false we have added the capability to do no transformation and write the XML as is into the variable. But this is not today's behavior - that is why prettyPrint defaults to true. 

The root cause is, that the transformer with property INDENT set only works as expected, if the input XML is not already pretty printed. I think this can even depend on the implementation of javax.xml.Transform on the classpath which is is used when transformerFactory.newTransformer(); is called.  So probably very hard to fix if at all. 

The formatter behavior is as follows: 
- If you put in an unformatted XML (everything in one line), the output is a formatted XML as expected
- if you put in an already formatted XML, the formatter adds an additional newline behind each line leading to blank lines. 
- if you put this output another time into the formatter you have already 2 blank lines after each line
- if the property INDENT is not set (prettyPrint = false) the output will not be touched

**Comment by ThorbenLindhauer** [comment]
Hey @jomu78,

Nikola is currently out of office, so I am taking over the PR review. It may take me a bit to get back to this however, since we are currently pretty busy with getting the 7.16 release out. I'll let you know once I had a closer look.

Cheers,
Thorben

File: dataformat-xml-dom/src/main/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriter.java
Comment (ThorbenLindhauer) [inline dataformat-xml-dom/src/main/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriter.java:91]: If this branch is reached, we will get a `NullPointerException` in `#getFormattingTransformer`. Can we have this method declare `throws ...` and then have exception handling only in `#getFormattingTransformer`?

File: dataformat-xml-dom/src/test/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriterTest.java
Comment (ThorbenLindhauer) [reply dataformat-xml-dom/src/test/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriterTest.java:15]: I disagree with this assessment, especially since disabled pretty printing will now also create a differently formatted XML as it no longer used the indent feature.

File: dataformat-xml-dom/src/test/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriterTest.java
Comment (ThorbenLindhauer) [inline dataformat-xml-dom/src/test/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriterTest.java:15]: * I have trouble understanding some of the logic of the tests in this class: for example why the "clones" are needed. 
* I don't think we need to test the lazy initialization as it's an implementation detail (that might change in the future). We should rather cover it with other tests.

Instead, I suggest to write tests with Spin's API, i.e. using `SpinXmlElement` instead of the the reader/writer classes. I wrote a test here that you could reuse: https://github.com/camunda/camunda-spin/commit/700f6e2103f7eee6a0d79f596a249bd5e4f9ca81

**Review by ThorbenLindhauer — CHANGES_REQUESTED** [review]
Hi @jomu78,

I'm sorry for the long delay with my review. I could only look into this last week and needed a bit to fully understand the motivation for this change (which I have documented in https://jira.camunda.com/browse/CAM-13699?focusedCommentId=210568&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-210568).

I have added some review notes that I want to ask you to address before we can merge the change. In addition, as this changes the behavior of XML serialization we should add a note to the 7.17 update guide which would include the updated Spin version. I will take care of that once your PR is merged.

Thanks for your patience and I hope you are still willing to get this through :)

Cheers,
Thorben

**Comment by ThorbenLindhauer** [comment]
I'm closing this PR due to inactivity, don't hesitate to reopen it if you want to continue.

**Comment by jomu78** [comment]
@ThorbenLindhauer : sorry for not responding, we have been quite busy the last weeks. But moved the ticket into our current sprint to get you some feedback asap.

**Comment by ThorbenLindhauer** [comment]
@jomu78 sounds great, I'll reopen the PR then. I'm going on holidays next week, and will be able to continue with you after Christmas. Cheers!

**Comment by jomu78** [comment]
@ThorbenLindhauer : I finally managed to work on the issue and have another commit for this pull request. Can you please open the PR again?

**Comment by ThorbenLindhauer** [comment]
@jomu78 sure, please push your branch :)

File: dataformat-xml-dom/src/main/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriter.java
Comment (jomu78) [reply dataformat-xml-dom/src/main/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriter.java]: removed formattingTemplates variable and synch, getFormattingTemplates called in getFormattingTransformer. I do not know what you mean / how to get this into the transformer factory as menttioned

File: dataformat-xml-dom/src/main/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriter.java
Comment (jomu78) [reply dataformat-xml-dom/src/main/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriter.java]: replaced by DomXmlDataFormatWriter.class.getClassLoader().getResourceAsStream(STRIP_SPACE_XSL)

File: dataformat-xml-dom/src/main/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriter.java
Comment (jomu78) [reply dataformat-xml-dom/src/main/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriter.java]: getFormattingTemplates now throws TransformerConfigurationException so handling is moved to getFormattingTransformer as requested

File: dataformat-xml-dom/src/test/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriterTest.java
Comment (jomu78) [reply dataformat-xml-dom/src/test/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriterTest.java:21]: completely rewritten the test on base of Spin as suggested in your commit. There is still one TODO marker in testDisabledPrettyPrintFormatted because it removes a blank line and I do not know why. On the other side, this is a test for a new feature as today's implementation always formats the XML. So it also does not break anything.

File: dataformat-xml-dom/src/test/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriterTest.java
Comment (jomu78) [reply dataformat-xml-dom/src/test/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriterTest.java:96]: agree, removed the TODO comment

File: dataformat-xml-dom/src/test/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriterTest.java
Comment (jomu78) [reply dataformat-xml-dom/src/test/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriterTest.java:21]: ok, removed the TODO

File: dataformat-xml-dom/src/test/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriterTest.java
Comment (jomu78) [reply dataformat-xml-dom/src/test/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriterTest.java:96]: Replaced with 
`DataFormat<SpinXmlElement> dataFormat = new DomXmlDataFormat(DataFormats.XML_DATAFORMAT_NAME);
SpinXmlElement spinXml = SpinFactory.INSTANCE.createSpin(xml, dataFormat);
`

File: dataformat-xml-dom/src/main/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriter.java
Comment (jomu78) [reply dataformat-xml-dom/src/main/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriter.java:93]: I tried to implement it the way you described.

File: dataformat-xml-dom/src/main/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriter.java
Comment (jomu78) [reply dataformat-xml-dom/src/main/java/org/camunda/spin/impl/xml/dom/format/DomXmlDataFormatWriter.java]: Fixed in last update

**Comment by ThorbenLindhauer** [comment]
Thanks for applying the review hints. I polished the changes a little more and merged them via https://github.com/camunda/camunda-spin/commit/c35641d40660c816656c17c0982946a1516e86e7. Next, I will proceed with building a new release of the Spin library, then update it in the https://github.com/camunda/camunda-bpm-platform repository and run our integration test suites against it.

Thanks again for your contribution and especially for your persistence with applying all the requested changes.

**Comment by tasso94** [comment]
I guess this contribution causes our JDK 8 tests to fail:
https://ci.cambpm.camunda.cloud/job/sideprojects/job/camunda-spin-UNIT-jdk-ibm-jdk-8-latest/

**Comment by ThorbenLindhauer** [comment]
Thanks for letting me know, I'll fix it via https://jira.camunda.com/browse/CAM-14806

**Pull request closed without merging**
